app/src/main/java/org/simple/clinic/registration/confirmpin/RegistrationConfirmPinUpdate.kt

```diff
@@ -16,20 +15,22 @@ class RegistrationConfirmPinUpdate : Update<RegistrationConfirmPinModel, Registr
       is RegistrationConfirmPinTextChanged -> next(model.withEnteredPinConfirmation(event.confirmPin))
       is RegistrationConfirmPinDoneClicked -> dispatch(ValidatePinConfirmation(model.enteredPinConfirmation, model.ongoingRegistrationEntry))
       is PinConfirmationValidated -> validateEnteredPin(model, event)
-      is RegistrationEntrySaved -> noChange()
+      is RegistrationEntrySaved -> dispatch(OpenFacilitySelectionScreen)
     }
   }
 
   private fun validateEnteredPin(
       model: RegistrationConfirmPinModel,
       event: PinConfirmationValidated
   ): Next<RegistrationConfirmPinModel, RegistrationConfirmPinEffect> {
-    val updatedModel = model.validatedPinConfirmation(event.result)
+    var updatedModel = model.validatedPinConfirmation(event.result)
 
-    return if (updatedModel.pinConfirmationDoesNotMatchEnteredPin) {
-      next(updatedModel, ClearPin as RegistrationConfirmPinEffect)
+    return if (updatedModel.pinConfirmationMatchesEnteredPin) {
+      updatedModel = updatedModel.updateRegistrationEntryWithPinConfirmation(updatedModel.enteredPinConfirmation, event.timestamp)
```

Comment:
Is it fine to mutate the model like this in the `Update` function?

Reply:
We are not actually mutating the model here, the model is still immutable. What are we doing is reassigning the reference of the variable `updatedModel` to a new instance of the model.

This is fine because the reference is local to this function. In addition, this is reassigned only in a single thread and will never involve concurrent access, so it's safe to do so.